Source/JavaScriptCore/dfg/DFGFixupPhase.cpp

```diff
@@ -1083,6 +1084,8 @@ class FixupPhase : public Phase {
             blessArrayOperation(node->child1(), node->child2(), node->child1()); // Rewrite child1 with ResolveRope.
             fixEdge<KnownStringUse>(node->child1());
             fixEdge<Int32Use>(node->child2());
+            if (node->op() == StringAt && m_graph.hasExitSite(node->origin.semantic, OutOfBounds))
```

Comment:
Fixup phase is after prediction propagation. So this is late. You need to check this status and update array mode in DFGByteCodeParser.